abi.html

```diff
@@ -2111,86 +2111,114 @@ <h4><a href="#vtable-ctor-entries"> 2.6.4 Construction Virtual Table entries </a
-    The standard guarantees (12.5 [class.free])
-    that this function will be passed the
-    number of bytes allocated with the previous array new expression.)
+<li>Otherwise, a cookie is required if:
 
-<p>
-<li> No cookie is required if the <code>new</code> operator being used
-    is <code>::operator new[](size_t, void*)</code>.
-
-<p>
-<li> Otherwise, this ABI requires a cookie, setup as follows:
   <ul>
-  <li> The cookie will have size <code>sizeof(size_t)</code>.
-  <li> Let <code>align</code> be the maximum alignment of
-  <code>size_t</code> and an element of the array to be allocated.
-  <li> Let <code>padding</code> be the maximum of
-  <code>sizeof(size_t)</code> and <code>align</code> bytes.
-  <li> The space allocated for the array will be the space required
-  by the array itself plus <code>padding</code> bytes.
-  <li> The alignment of the space allocated for the array will be
-  <code>align</code> bytes.
-  <li> The array data will begin at an offset of <code>align</code> bytes
-  from the space allocated for the array.
-  <li> The cookie will be stored in the <code>sizeof(size_t)</code> bytes
-  immediately preceding the array data.
+    <li><code>T</code> has a non-trivial destructor, or else
+
+    <li>there is an unambiguous usual array deallocation function
+        for <code>T</code> (as would be found for the <code>delete[]</code>
+        operator on an operand of type <code>T*</code>), and that
+        function takes a <code>size_t</code> argument.
+
+      <p>This operator must be looked up as if for a <code>delete[]</code>
+        without a <code>::</code> prefix, even if the <code>new</code>
+        expression does include a <code>::</code> prefix.  This is
+        necessary to ensure consistency between <code>new</code>
+        and <code>delete</code>, as it is not always ill-defined
+        under the standard to mix the use of global and class-scoped
+        allocation and deallocation functions.  (For example, a
+        class-scoped deallocation function might delegate to the
+        corresponding global deallocation function.)
+    </ul>
   </ul>
 
-  <i>
-  <p>
-  These rules have the following consequences:
-  <ul>
-  <li> The array elements and the cookie are all aligned naturally.
-  <li> Padding will be required if <code>sizeof(size_t)</code>
-  is smaller than the array element alignment,
-  and if present will precede the cookie.
-  </ul>
-  </i>
+<p>If a cookie is required, the allocation is adjusted as follows:
 
+<ul>
+  <li>Let <code>cookie_t</code> be the type of a cookie.  Under the
+    standard ABI, this is <code>size_t</code>; however, some platforms
+    use a variant ABI which stores additional information.
+
+  <li>Let <code>count</code> be the total number of objects of type
+    <code>T</code> that will be created, scaled appropriately by the
+    bounds of any nested arrays.  This is the value that will be stored
```

Comment:
The "scaled appropriately" part seems redundant/tautological. At first glance I thought it almost looks like it's saying to do that scaling twice (as in "let `count` be the total number of elements multiplied by the array bounds").

What do you think about "Let `count` be the total number of objects of type `T` that will be created, as determined by the bounds of any nested arrays." instead?

Reply:
Updated